diff --git a/rxinitfs b/rxinitfs
index 6b4f632..3afd87c 100755
--- a/rxinitfs
+++ b/rxinitfs
@@ -3,12 +3,31 @@
 set -e
 
 find_mods() {
-  find "@LIBDIR@/modules/$(uname -r)/" -name "$1.ko.xz" -print0 -exec cp {} "$2/lib/modules-load/"
+  local f=$(find "@LIBDIR@/modules/" -name "$1.ko.xz")
+  [[ -z "$f" ]] && (>&2 echo "Cannot find module $1") && exit 1
+  cp "$f" "$2/lib/modules-load/"
   unxz "$2/lib/modules-load/$1.ko.xz"
 
-  for d in $(modinfo "$2/lib/modules-load/$1.ko.xz" | grep depends | cut -d ':' -f2 | xargs | tr ',' '\n'); do
+  local kernel=$(echo "$f" | cut -d '/' -f4)
+  if echo "@LIBDIR@" | grep "usr/lib" >/dev/null; then
+    kernel=$(echo "$f" | cut -d '/' -f5)
+  fi
+  local f2=$(echo "$f" | sed "s|/lib/modules/$kernel/||g")
+
+  local deps=()
+  for d in $(modinfo "$2/lib/modules-load/$1.ko" | grep depends | cut -d ':' -f2 | xargs | tr ',' '\n'); do
     find_mods "$d" "$2"
+    local f3=$(find "@LIBDIR@/modules/" -name "$d.ko.xz")
+    local kernel2=$(echo "$f3" | cut -d '/' -f4)
+    if echo "@LIBDIR@" | grep "usr/lib" >/dev/null; then
+      kernel2=$(echo "$f3" | cut -d '/' -f5)
+    fi
+    deps+=($(echo "$f" | sed "s|/lib/modules/$kernel2/||g"))
   done
+
+  mkdir -p "$2$(dirname $f)"
+  cp "$f" "$2$f"
+  echo "$f2: ${deps[@]}" >> "$2/lib/modules/$kernel/modules.dep"
 }
 
 build_initramfs() {
@@ -20,7 +39,14 @@ build_initramfs() {
   cp "@SYSCONFDIR@/rxinitfs.conf" etc/rxinitfs.conf
   source etc/rxinitfs.conf
   for m in $MODULES; do
-    find_mods "$m" "$tempd"
+    if echo "$m" | grep "\*" >/dev/null; then
+      local mods=($(find "@LIBDIR@/modules" -name "$m.ko.xz" | awk -F/ "{print \$(NF)}" | cut -d '.' -f1))
+      for m2 in "${mods[@]}"; do
+        find_mods "$m2" "$tempd"
+      done
+    else
+      find_mods "$m" "$tempd"
+    fi
   done
   find . | cpio -H newc -o > "$tempcpio"
   cd ..
diff --git a/template/etc/group b/template/etc/group
index 18acc30..c5bb679 100644
--- a/template/etc/group
+++ b/template/etc/group
@@ -1 +1,6 @@
 root:x:0:root
+tty:x:5:
+uucp:x:987:
+lp:x:991:
+disk:x:994:
+kmem:x:997:
diff --git a/template/init b/template/init
index 1998314..6ac953f 100755
--- a/template/init
+++ b/template/init
@@ -30,7 +30,13 @@ mknod /dev/tty c 5 0
 source /etc/rxinitfs.conf
 echo "Loading modules"
 for m in $MODULES; do
-  loadmod $m
+  if echo "$m" | grep "\*" >/dev/null; then
+    for m2 in $(find "/lib/modules" -name "$m.ko.xz" | awk -F/ "{print \$(NF)}" | cut -d '.' -f1); do
+      loadmod "$m2"
+    done
+  else
+    loadmod "$m"
+  fi
 done
 
 mdev -s
