diff --git a/rxinitfs b/rxinitfs
index 3b010f8..268864a 100755
--- a/rxinitfs
+++ b/rxinitfs
@@ -6,7 +6,7 @@ build_initramfs() {
   local outpath=$1
   local tempd=$(mktemp -d)
   local tempcpio=$(mktemp)
-  cp "@PREFIX@/share/rxinitfs/template/" "$tempd"
+  cp "@PREFIX@/share/rxinitfs/template/"* "$tempd"
   cd "$tempd"
   cp "@LIBDIR@/modules/"* usr/lib/modules/ -r
   find . | cpio -H newc -o > "$tempcpio"
